Solver/utilities.py

```diff
@@ -0,0 +1,202 @@
+            material_parameters, mat_param_ref, l_ref)
+
+    if u_ana is not None:
+        u_ana_dl = get_dimensionless_displacement(u_ana, l_ref, u_ref)
+        return domain_dl, boundary_conditions_dl, body_forces_dl, \
+            material_parameters_dl, u_ana_dl
+
+    else:
+        return domain_dl, boundary_conditions_dl, body_forces_dl, \
+            material_parameters_dl
+
+
+def get_dimensional_displacement(u_hat_dimless, boundary_conditions,
+                                 u_ref, l_ref):
+    dim = len(u_hat_dimless)
+    N = [u_hat_dimless.function_space().spaces[0].bases[i].N
+         for i in range(dim)]
+
+    domain_dimless = [u_hat_dimless.function_space().spaces[i].bases[i].domain
+                      for i in range(dim)]
+    domain = [
+            [
+                domain_dimless[i][j] * l_ref for j in range(2)
+                ]
+            for i in range(dim)
+        ]
+
+    # vector space for solution
+    vec_space = []
+    for i in range(dim):  # nb of displacement components
+        tens_space = []
+        for j in range(dim):  # nb of FunctionSpaces for each component
+            basis = sf.FunctionSpace(N[j], domain=domain[j], family='legendre',
+                                     bc=boundary_conditions[i][j])
+            tens_space.append(basis)
+        vec_space.append(sf.TensorProductSpace(sf.comm, tuple(tens_space)))
+
+    V = sf.VectorSpace(vec_space)
+
+    # actual solution (same coefficients, different vector space)
+    u_hat = sf.Function(V)
+
+    for i in range(dim):
+        # u has the same expansions coefficients as u_dimless
+        u_hat[i] = u_ref * u_hat_dimless[i]
+
+    return u_hat
+
+
+def get_dimensionless_domain(domain, l_ref):
+    return tuple(np.array(domain) / l_ref)
+
+
+def get_dimensionless_material_parameters(material_parameters,
+                                          mat_param_ref, l_ref):
+    if len(material_parameters) == 2:  # cauchy elasticity
+        return tuple(np.array(material_parameters) / mat_param_ref)
+    elif len(material_parameters) == 7:  # gradient elasticity
+        lame_parameters = np.array(
+                material_parameters
+                )[[0, 1]] / mat_param_ref
+        gradient_parameters = np.array(material_parameters)[2:] \
+            / mat_param_ref / l_ref ** 2
+        # wrap material parameters in a tuple
+        return tuple(np.hstack((lame_parameters, gradient_parameters)))
+
+
+def get_dimensionless_boundary_conditions(boundary_conditions, l_ref,
+                                          u_ref):
+    # assign bc type to each boundary condition
+    def assign_bc_type(bc):
+        if bc is None:
+            return 'ZERO_TRACTION'
+        elif isinstance(bc, tuple):  # dirichlet-bcs are given as tuple
+            return 'DIRICHLET'
+        elif isinstance(bc, dict):  # all bcs can be defined via a dictionary
+            return 'ARBITRARY'
+        else:
+            raise NotImplementedError()
```

Comment:
Should we use an `enum` to handle the different boundary conditions? This might be a better than using strings.

Reply:
I agree. I will implement this in the next commit.

Reply:
Do you mean on a separate branch?

Reply:
No, I will implement it on this branch. In one of the upcoming commits ... 😄